### PR TITLE
Reland "[X86][AVX10.2] Support AVX10.2 option and VMPSADBW/VADDP[D,H,S] new instructions (#101452)"

### PR DESCRIPTION
**Pull request by phoebewang**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/828965

### CODE REVIEW

File: llvm/utils/TableGen/X86RecognizableInstr.cpp [hunk 1]
```diff
@@ -223,6 +227,10 @@ InstructionContext RecognizableInstr::insnContext() const {
           insnContext = EVEX_KB_U(IC_EVEX_XD);
         else if (OpPrefix == X86Local::PS)
           insnContext = EVEX_KB_U(IC_EVEX);
+        else {
+          errs() << "Instruction does not use a prefix: " << Name << "\n";
+          llvm_unreachable("Invalid prefix");
```
Comment (KanRobert) [inline llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: The default op prefix for `EVEX` is `PS` in compiler. 6e20df1a3b0f7654b2821fe182c7ae9bd52672e6

Should we use the message like

```
llvm_unreachable(("Invalid prefix for" + Name).c_str())
```

File: llvm/utils/TableGen/X86RecognizableInstr.cpp [hunk 2]
```diff
@@ -214,6 +214,10 @@ InstructionContext RecognizableInstr::insnContext() const {
           insnContext = EVEX_KB_U(IC_EVEX_W_XD);
         else if (OpPrefix == X86Local::PS)
           insnContext = EVEX_KB_U(IC_EVEX_W);
+        else {
+          errs() << "Instruction does not use a prefix: " << Name << "\n";
```
Comment (KanRobert) [inline llvm/utils/TableGen/X86RecognizableInstr.cpp:218]: ditto

**Comment by KanRobert** [comment]
Is this fail related to your PR?

```
LLVM-Unit :: tools/llvm-cfi-verify/./CFIVerifyTests/2/49
```

File: llvm/utils/TableGen/X86RecognizableInstr.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: Do you mean remove the `errs() << ...`? How about the others? Both VEX and EVEX have the 4 prefixes.

**Comment by phoebewang** [comment]
> Is this fail related to your PR?
> 
> ```
> LLVM-Unit :: tools/llvm-cfi-verify/./CFIVerifyTests/2/49
> ```

I think so, investigating..

File: llvm/utils/TableGen/X86RecognizableInstr.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: I meant there is always a default prefix opcode `PS` for EVEX prefix, so `does not use a prefix` is misleading.

Yes, I suggest merging the `errs() << ` into `llvm_unreachable`.

Comment (e-kud) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: Oh, I've asked myself some time ago if there is a difference between `llvm_unreachable` and `errs()`. The deal is in production build. The message from `llvm_unreachable` is not shown.

**Review by e-kud — APPROVED** [review]
LGTM

File: llvm/utils/TableGen/X86RecognizableInstr.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: I think if the `errs()` is followed by `llvm_unreachable`, the message will be optimized away too in prod build.

File: compiler-rt/lib/builtins/cpu_model/x86.c [hunk 3]
```diff
@@ -978,8 +978,20 @@ static void getAvailableFeatures(unsigned ECX, unsigned EDX, unsigned MaxLeaf,
...
-    setFeature(FEATURE_AVX10_1_512);
+  if (HasLeaf7Subleaf1 && ((EDX >> 19) & 1) && HasLeaf24) {
+    bool Has512Len = (EBX >> 18) & 1;
+    int AVX10Ver = EBX & 0xff;
+    if (AVX10Ver >= 2) {
+      setFeature(FEATURE_AVX10_2_256);
+      if (Has512Len)
+        setFeature(FEATURE_AVX10_2_512);
+    }
+    if (AVX10Ver >= 1) {
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply compiler-rt/lib/builtins/cpu_model/x86.c:989]: No. There're two problems from compiling:

- compiler-rt doesn't support `llvm_unreachable`;
- some buildbot will warn about `[[fallthrough]]`

But yes, I found the implications problem during revisiting it, i.e., an old version compiler-rt runs on future HW.

File: llvm/utils/TableGen/X86RecognizableInstr.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: No sure about the iostream operations, but I remember `llvm_unreachable` won't optimizate function call like printf.

Comment (phoebewang) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: So I lean toward to keep both.

**Comment by phoebewang** [comment]
> > Is this fail related to your PR?
> > ```
> > LLVM-Unit :: tools/llvm-cfi-verify/./CFIVerifyTests/2/49
> > ```
> 
> I think so, investigating..

This is an interesting issue. The problem is due to https://github.com/llvm/llvm-project/blob/main/llvm/unittests/tools/llvm-cfi-verify/FileAnalysis.cpp#L242, which is an illegal encoding. While I switched the `!readModRM(insn) && isNF(insn)` to reuse the parser of ModRM. But the parsering or not affects the length of the illegal instruction and results the fail parser of the expected `0x0f, 0x0b // 28: ud2`, hence the failure.

File: llvm/utils/TableGen/X86RecognizableInstr.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: Update the error message?

**Review by KanRobert — APPROVED** [review]
LGTM with one minor suggestion

File: llvm/utils/TableGen/X86RecognizableInstr.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/utils/TableGen/X86RecognizableInstr.cpp:232]: There are only 2 out of 4141 `grep -rwn llvm_unreachable llvm/lib/ | wc` use `+` to show more detail. I think `llvm_unreachable` is mainly used to show the position to developer, so no need detailed information.

**Pull request merged**